src/progress.rs

```diff
@@ -47,118 +47,187 @@ impl Default for ProgressState {
     }
 }
 
+#[derive(Clone, Debug, Default)]
+struct Configuration {
+    voters: FxHashSet<u64>,
+    learners: FxHashSet<u64>,
+}
+
 /// `ProgressSet` contains several `Progress`es,
 /// which could be `Leader`, `Follower` and `Learner`.
 #[derive(Default, Clone)]
 pub struct ProgressSet {
-    voters: FxHashMap<u64, Progress>,
-    learners: FxHashMap<u64, Progress>,
+    progress: FxHashMap<u64, Progress>,
+    configuration: Configuration,
 }
 
 impl ProgressSet {
     /// Creates a new ProgressSet.
-    pub fn new(voter_size: usize, learner_size: usize) -> Self {
+    pub fn new() -> Self {
         ProgressSet {
-            voters: HashMap::with_capacity_and_hasher(voter_size, Default::default()),
-            learners: HashMap::with_capacity_and_hasher(learner_size, Default::default()),
+            progress: Default::default(),
+            configuration: Default::default(),
+        }
+    }
+
+    /// Create a progress sete with the specified sizes already reserved.
+    pub fn with_capacity(voters: usize, learners: usize) -> Self {
+        ProgressSet {
+            progress: HashMap::with_capacity_and_hasher(
+                voters + learners,
+                FxBuildHasher::default(),
+            ),
+            configuration: Configuration {
+                voters: HashSet::with_capacity_and_hasher(voters, FxBuildHasher::default()),
+                learners: HashSet::with_capacity_and_hasher(learners, FxBuildHasher::default()),
+            },
         }
     }
 
     /// Returns the status of voters.
-    pub fn voters(&self) -> &FxHashMap<u64, Progress> {
-        &self.voters
+    #[inline]
+    pub fn voters(&self) -> impl Iterator<Item = (&u64, &Progress)> {
+        self.progress
+            .iter()
+            .filter(move |(&k, _)| self.has_voter(k))
     }
 
     /// Returns the status of learners.
-    pub fn learners(&self) -> &FxHashMap<u64, Progress> {
-        &self.learners
+    #[inline]
+    pub fn learners(&self) -> impl Iterator<Item = (&u64, &Progress)> {
+        self.progress
+            .iter()
+            .filter(move |(&k, _)| self.has_learner(k))
+    }
+
+    /// Returns the mutable status of voters.
+    #[inline]
+    pub fn voters_mut(&mut self) -> impl Iterator<Item = (&u64, &mut Progress)> {
+        let set = self.voter_ids().cloned().collect::<FxHashSet<_>>();
```

Comment:
You can hint the compiler by borrow progress explicitly like:
```
let progress = &mut self.progress;
self.voter_ids().map(|id| (*id, progress.get_mut(id)))
```

Reply:
Due to lifetime requirements this doesn't work.

Since we are returning an iterator and the `let progress = &mut self.progress;` drops off at the end of the function. The iterator needs that borrow to survive the entire time.

Example:

```rust
    /// Returns the mutable status of voters.
    #[inline]
    pub fn voters_mut(&mut self) -> impl ExactSizeIterator<Item = (u64, &mut Progress)> {
        let progress = &mut self.progress;
        self.voter_ids().iter().map(|id| (*id, progress.get_mut(id).unwrap()))
    }
```

Results in:

```rust
error[E0495]: cannot infer an appropriate lifetime for autoref due to conflicting requirements
   --> src/progress.rs:105:57
    |
105 |         self.voter_ids().iter().map(|id| (*id, progress.get_mut(id).unwrap()))
    |                                                         ^^^^^^^
    |
note: first, the lifetime cannot outlive the lifetime  as defined on the body at 105:37...
   --> src/progress.rs:105:37
    |
105 |         self.voter_ids().iter().map(|id| (*id, progress.get_mut(id).unwrap()))
    |                                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
note: ...so that reference does not outlive borrowed content
   --> src/progress.rs:105:48
    |
105 |         self.voter_ids().iter().map(|id| (*id, progress.get_mut(id).unwrap()))
    |                                                ^^^^^^^^
note: but, the lifetime must be valid for the anonymous lifetime #1 defined on the method body at 103:5...
   --> src/progress.rs:103:5
    |
103 | /     pub fn voters_mut(&mut self) -> impl ExactSizeIterator<Item = (u64, &mut Progress)> {
104 | |         let progress = &mut self.progress;
105 | |         self.voter_ids().iter().map(|id| (*id, progress.get_mut(id).unwrap()))
106 | |     }
    | |_____^
    = note: ...so that the types are compatible:
            expected (u64, &mut progress::Progress)
               found (u64, &mut progress::Progress)
```

I was able to avoid this with the implementation given because we're cloning a copy of the data and allowing the iterator to capture it. We could have cloned `Progress` to have something closer to what you suggested but that's significantly more expensive, so I settled for cloning the `HashSet`.:

```rust
pub fn voters_mut(&mut self) -> impl Iterator<Item = (&u64, &mut Progress)> {
       let set = self.voter_ids().clone();
       self.progress
           .iter_mut()
           .filter(move |(&k, _)| set.contains(&k));
}
```

Reply:
The problem is due to `get_mut` instead of the explicit borrow. I revert the order to make a short travel. Since compiler is not smart enough, keep the original order should be OK.

```
let ids = &self.configuration.voters;
self.progress.iter_mut().filter(move |(k, _)| ids.contains(k)) 
```

Reply:
Any updates?